micov/_io.py

```diff
@@ -265,13 +265,19 @@ def _reader(sam):
     elif isinstance(sam, io.BytesIO):
         yield sam
     else:
-        with lzma.open(sam) as fp:
-            data = fp.read()
-            yield data
+        if sam.endswith('.sam.xz'):
+            yield lzma.open(sam, mode='rt', encoding='utf-8')
+        elif sam.endswith('.sam.gz'):
+            yield gzip.open(sam, mode='rt', encoding='utf-8')
+        else:
+            yield open(sam, 'r')
 
 
 def _buf_to_bytes(buf):
-    return io.BytesIO(b''.join(buf))
+    if isinstance(buf[0], str):
+        return io.BytesIO(b''.join(s.encode() for s in buf))
+    else:
+        return io.BytesIO(b''.join(buf))
```

Comment:
This looks inverted. If `buf[0]` is `str` then it is already `utf-8` by definition. Also, joining `bytes` (e.g., `b''.join(blah)`) will only work if `blah` is also `bytes`. So if the data are `utf-8` then the join should be `''.join(blah)`